contracts/cw-proposal-single/src/contract.rs

```diff
@@ -338,6 +369,52 @@ pub fn execute_update_config(
         .add_attribute("sender", info.sender))
 }
 
+pub fn execute_add_hook(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    address: String,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+    if config.dao != info.sender {
+        // Only DAO can add hooks
+        return Err(ContractError::Unauthorized {});
+    }
+
+    let validated_address = deps.api.addr_validate(&address)?;
+
+    HOOKS
+        .add_hook(deps.storage, validated_address.clone())
+        .map_err(ContractError::HookError)?;
+
+    Ok(Response::default()
+        .add_attribute("action", "add_hook")
+        .add_attribute("address", validated_address))
+}
+
+pub fn execute_remove_hook(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    address: String,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+    if config.dao != info.sender {
+        // Only DAO can remove hooks
+        return Err(ContractError::Unauthorized {});
+    }
+
+    let validated_address = deps.api.addr_validate(&address)?;
+
+    HOOKS
+        .remove_hook(deps.storage, validated_address.clone())
```

Comment:
Same as above about losing the clone by using `address` below. Please ignore if you feel the code is less readable that way.